### PR TITLE
Add CSV loading commands

### PR DESCRIPTION
**Pull request by queryproc**

### CODE REVIEW

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        };
+    }
+
+    private void loadVertices() throws IOException {
+        List<String> verticesData = Files.readAllLines(Paths.get(structuredQuery.getFilePath()),
+            StandardCharsets.UTF_8);
+
+        int idIndex = -1;
+        int typeIndex = -1;
+        String[] header = verticesData.get(0).split(structuredQuery.getCsvSeparator());
```
Comment (chathura-g21) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java]: trim here before splitting? An extraneous space can make the iteration below go wrong

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+                } catch (Exception e) {
+                    return new Message(String.format("Error loading graph " + dataLoadedType +
+                        " from '%s'. Either a problem with the file's path, the structure or " +
+                        "content.", structuredQuery.getFilePath()), true /* isError */);
+                }
+            }
+        };
+    }
+
+    private void loadVertices() throws IOException {
```
Comment (chathura-g21) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java]: Can you give an explanation of the formats you are expecting in a javadoc here?

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+                String PropertyDataType = columnSplit.length < 2 ?  DataType.STRING.name() :
+                    columnSplit[1];
+                Pair<Short, DataType> propertyShortAndDataType = TypeAndPropertyKeyStore.
+                    getInstance().mapStringPropertyKeyValueToShortAndDataTypeOrInsert(propertyName,
+                    PropertyDataType);
+                vertexPropertiesIndexMap.put(i, propertyShortAndDataType);
+            }
+        }
+
+        if (typeIndex == -1 && null == structuredQuery.getCsvDataType()) {
```
Comment (chathura-g21) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java]: Be consistent on whether the constant is in the LHS or RHS here. I think our convention is LHS.

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    }
+
+    private void loadEdges() throws IOException {
+        List<String> edgesData = Files.readAllLines(Paths.get(structuredQuery.getFilePath()),
+            StandardCharsets.UTF_8);
+
+        int fromVertexIndex = -1;
+        int toVertexIndex = -1;
+        int typeIndex = -1;
+        String[] header = edgesData.get(0).split(structuredQuery.getCsvSeparator());
```
Comment (chathura-g21) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java]: trim.

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            } else {
+                String[] propertyNameAndType = header[i].split(COLON, 2);
+                Pair<Short, DataType> propertyShortAndDataType = TypeAndPropertyKeyStore.
+                    getInstance().mapStringPropertyKeyValueToShortAndDataTypeOrInsert(
+                        propertyNameAndType[0], propertyNameAndType[1]);
+                edgePropertiesIndexMap.put(i, propertyShortAndDataType);
+            }
+        }
+
+        if (typeIndex == -1 && null == structuredQuery.getCsvDataType()) {
```
Comment (chathura-g21) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java]: -1 == typeIndex

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+                forwardLists[i] = new SortedAdjacencyList();
+                backwardLists[i] = new SortedAdjacencyList();
+            }
+        }
+        Graph.getInstance().setForwardAdjLists(forwardLists);
+        Graph.getInstance().setBackwardAdjLists(backwardLists);
+        Graph.getInstance().setHighestPermanentAndMergedVertexId(maxVertexId);
+    }
+
+    private void loadEdges() throws IOException {
```
Comment (chathura-g21) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java]: Javadoc with expected format please.

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        List<String> edgesData = Files.readAllLines(Paths.get(structuredQuery.getFilePath()),
```
Comment (chathura-g21) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java]: Edge lists can easily not fit in memory. Why not a buffered reader?

File: src/test/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVPlannerTest.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Tests the {@link LoadDataFromCSVQueryPlanner}.
+ *
+ * Resource files used in these tests are from the train benchmark project.
+ * https://github.com/FTSRG/trainbenchmark/tree/graphflow/models
+ */
+public class LoadDataFromCSVPlannerTest {
```
Comment (queryproc) [inline src/test/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVPlannerTest.java]: Everything looks good to me in the test.
I expect you to add testing like you did so far for each edge in the schema image I gave you.
Can you actually add that image under test/resources and mention it in this class doc.

Also for the resources, change your dir names to edge_data_csv and vertices_data_csv. The image can go directly under resources or add a schema somewhere in there.

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    /**
+     * Loads vertices from a CSV file into the Graph.
+     * The header is of the form `id:ID,[:LABEL],[property:TYPE]` (in any order), and the rows
+     * contain the corresponding data. The TYPE is STRING by default, and mentioning `id` means
+     * adding it as a property of type INT.
+     *
+     * @throws IOException if I/O errors happen given the file path provided by the user.
+     */
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java]: between that big description and the @throws add different examples of headers and rows and what the code does.
I expect you to mention that having something to the left of :ID means it will be added as property.
mention that w.e. to the left of label is ignored.
Just give examples of rows for completeness. Do the same for edges.

File: src/main/java/ca/waterloo/dsg/graphflow/query/parser/GraphflowVisitor.java [hunk 10]
```diff
@@ -98,6 +98,39 @@ public AbstractStructuredQuery visitDurabilityQuery(DurabilityQueryContext ctx)
...
+    public AbstractStructuredQuery visitLoadCSVQuery(LoadCSVQueryContext ctx) {
+        StructuredQuery structuredQuery = new StructuredQuery();
+        if (null != ctx.loadEdgesPattern()) {
+            visitLoadEdgesPattern(ctx.loadEdgesPattern(), structuredQuery);
+        } else {
+            visitLoadVerticesPattern(ctx.loadVerticesPattern(), structuredQuery);
+        }
+        structuredQuery.setFilePath(getUnquotedString(ctx.stringLiteral().getText()));
+        String separator = getUnquotedString(ctx.QuotedCharacter().getText());
+        if (separator.equals("|")) {
```
Comment (queryproc) [inline src/main/java/ca/waterloo/dsg/graphflow/query/parser/GraphflowVisitor.java]: Make an issue about potential delimiters that need to be escaped like this.

File: src/test/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVPlannerTest.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    @Test
+    public void testFindNumberOfTargets() {
+        String matchQuery = "MATCH (a)-[:target]->(b) return COUNT(*);";
+        QueryResult queryResult = new OneTimeMatchQueryPlanner(new StructuredQueryParser().parse(
+            matchQuery)).plan().execute();
+        int edgeCount = Integer.parseInt((String) ((Tuples) queryResult).getTuples().get(0)[0]);
+        Assert.assertEquals(edgeCount, EXPETED_TARGET_EDGES);
+    }
```
Comment (queryproc) [inline src/test/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVPlannerTest.java]: We agreed there would be tests without the edge type but relying on the vertex filters and you would sum them up and assert on the equality on the two matches.

File: src/main/java/ca/waterloo/dsg/graphflow/query/parser/GraphflowVisitor.java (hunk 10 above)
Comment (nare469) [reply src/main/java/ca/waterloo/dsg/graphflow/query/parser/GraphflowVisitor.java]: done https://github.com/graphflow/graphflow/issues/35

File: src/test/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVPlannerTest.java (hunk 8 above)
Comment (nare469) [reply src/test/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVPlannerTest.java]: done

File: src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java (hunk 9 above)
Comment (nare469) [reply src/main/java/ca/waterloo/dsg/graphflow/query/planner/LoadDataFromCSVQueryPlanner.java]: done
